cw/capture.py

```diff
@@ -1167,118 +1169,266 @@ def capture_otbn_vertical(ot, ktp, fw_bin, pll_frequency, capture_cfg, device_cf
     # Initialize some curve-dependent parameters.
     if capture_cfg["curve"] == 'p256':
         curve_order_n = 0xffffffff00000000ffffffffffffffffbce6faada7179e84f3b9cac2fc632551
-        # key_bytes = 256 // 8
+        key_bytes = 256 // 8
         seed_bytes = 320 // 8
+        modinv_share_bytes = 320 // 8
+        modinv_mask_bytes = 128 // 8
     else:
         # TODO: add support for P384
-        raise NotImplementedError(f'Curve {capture_cfg["curve"]} is not supported')
-
-    # Check the lengths in the key/plaintext generator. In this case, "key"
-    # means seed and "plaintext" means mask.
-    if ktp.keyLen() != seed_bytes:
-        raise ValueError(f'Unexpected seed length: {ktp.keyLen()}.\n'
-                         f'Hint: set key len={seed_bytes} in the configuration file.')
-    if ktp.textLen() != seed_bytes:
-        raise ValueError(f'Unexpected mask length: {ktp.textLen()}.\n'
-                         f'Hint: set plaintext len={seed_bytes} in the configuration file.')
-
-    # Seed the RNG and generate a random fixed seed for all traces of the
-    # keygen operation.
-    seed_fixed = ktp.next_key()
-    print(f'fixed seed = {seed_fixed.hex()}')
-    if len(seed_fixed) != seed_bytes:
-        raise ValueError(f'Fixed seed length is {len(seed_fixed)}, expected {seed_bytes}')
-
-    # Expected key is `seed mod n`, where n is the order of the curve and
-    # `seed` is interpreted as little-endian.
-    expected_fixed_key = int.from_bytes(seed_fixed, byteorder='little') % curve_order_n
+        raise NotImplementedError(
+            f'Curve {capture_cfg["curve"]} is not supported')
 
     # register ctrl-c handler to not lose already recorded traces if measurement is aborted
     signal.signal(signal.SIGINT, partial(abort_handler, project))
 
-    sample_fixed = 1
-    # Loop to collect each power trace
-    for _ in tqdm(range(capture_cfg["num_traces"]), desc='Capturing',
-                  ncols=80):
+    if capture_cfg["app"] == 'keygen':
+        # Check the lengths in the key/plaintext generator. In this case, "key"
+        # means seed and "plaintext" means mask.
+        if ktp.keyLen() != seed_bytes:
+            raise ValueError(
+                f'Unexpected seed length: {ktp.keyLen()}.\n'
+                f'Hint: set key len={seed_bytes} in the configuration file.')
+        if ktp.textLen() != seed_bytes:
+            raise ValueError(
+                f'Unexpected mask length: {ktp.textLen()}.\n'
+                f'Hint: set plaintext len={seed_bytes} in the configuration file.'
+            )
 
-        if reset_firmware:
-            ot.program_target(fw_bin, pll_frequency)
+        # select the otbn app on the device (0 -> keygen, 1 -> modinv)
+        ot.target.simpleserial_write("a", bytearray([0x00]))
+        time.sleep(0.3)
+
+        # Seed the RNG and generate a random fixed seed for all traces of the
```

Comment:
Aren't we need to seed RNG before generating a fixed seed value?

Reply:
Yes of course!
Don't know why this is missing here. I'll fix it!!
Thanks for pointing this out!